pkg/kube-metrics/metrics.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	ksmetric "k8s.io/kube-state-metrics/pkg/metric"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("kubemetrics")
+
+// ServeCRMetrics generates CustomResource specific metrics based on the inputted custom resource GVK.
+// A list of namespaces, ns, can be passed to ServeCRMetrics to scope the generated metrics. Passing nil or
+// an empty list of namespaces will result in an error.
+// The function also starts serving the generated collections of the metrics on given host and port.
+func ServeCRMetrics(cfg *rest.Config,
+	ns []string,
+	operatorGVKs []schema.GroupVersionKind,
+	host string, port int32) error {
+	// We have to have at least one namespace.
+	if len(ns) < 1 {
+		return errors.New("namespaces were empty; pass at least one namespace to generate custom resource metrics")
+	}
+	// Create new unstructured client.
+	var collectors [][]kcollector.Collector
+	log.V(1).Info("Starting collecting operator types")
+	// Loop through all the possible operator/custom resource specific types.
+	for _, gvk := range operatorGVKs {
+		api := gvk.GroupVersion().String()
+		kind := gvk.Kind
+		// Generate metric based on the kind.
+		metricFamilies := generateMetricFamilies(gvk.Kind)
+		log.V(1).Info("Generating metric families", "apiVersion", api, "kind", kind)
+		dclient, err := newClientForGVK(cfg, api, kind)
+		if err != nil {
+			return err
+		}
+		// Generate collector based on the group/version, kind and the metric families.
+		c := NewCollectors(dclient, ns, api, kind, metricFamilies)
+		collectors = append(collectors, c)
+	}
+	// Start serving metrics.
+	log.V(1).Info("Starting serving custom resource metrics")
+	go ServeMetrics(collectors, host, port)
+
+	return nil
+}
+
+func generateMetricFamilies(kind string) []ksmetric.FamilyGenerator {
+	helpText := fmt.Sprintf("Information about the %s operator custom resource replica.", kind)
```

Comment:
This metrics help text is per CR type right? So for an operator with 2 CR types/Kinds, e.g Prometheus and ServiceMonitor, this will show up as:
```
Information about the Prometheus operator custom resource replica
Information about the ServiceMonitor operator custom resource replica
```

So probably drop the word `operator`. And `replica` too. Not sure what that's meant to convey.
```suggestion
	helpText := fmt.Sprintf("Information about the %s custom resource .", kind)
```

Reply:
Also looking below at the metrics generated for this CR kind, the metric seems to be the count of the number of instances of a particular CR kind?
For e.g I have two named instances `app1` `app2` of my CR type `Memcached`, will the count be 2?

Reply:
Yes below is the example of how that will look like:
```
# HELP jonthebeach_info Information about the JOnTheBeach operator custom resource replica.
# TYPE jonthebeach_info gauge
jonthebeach_info{namespace="default",jonthebeach="example-jonthebeach"} 1
jonthebeach_info{namespace="default",jonthebeach="example-jonthebeach2"} 1
```
So if you select by CR you will get the correct count.

Before the above change hence the operator in the name, but you get the idea :)